[BUGFIX] Sustains now give consistent scores [+ songScore as Float]

xenkap opened the pull request:
<!-- Please check for duplicates or similar PRs before submitting this PR. -->
## Does this PR close any issues? If so, link them below.
N/A

## Briefly describe the issue(s) fixed.
Sustain notes previously did not consistently give score-- it was all dependent on how long you held it for, regardless of when you started holding, disconnected from the actual length of the note. This caused sustains hit earlier to reward more points than sustains hit later, as well as other undesirable effects. This behavior is now fixed.

These demonstrations are done with a script that disables regular note scoring, leaving only sustain scores to count.

Before: 

https://github.com/user-attachments/assets/6882c412-0b71-43a2-9f9e-3b5911b04122

After (VIDEO SLIGHTLY OUTDATED, 100% FIXED NOW): 

https://github.com/user-attachments/assets/96338947-9574-4bd8-b216-824e5288b921


This PR places scoring behavior in Strumline, which may be undesirable.
An alternative approach would be to add a variable in SustainTrail.hx that tracks PlayState's last read on it, and then performing logic where it used to be done anyway. I tried this, but it didn't exactly work out...

This PR also changes songScore to be a Float, and all score-related scripting variables to Float as well.
All saving and displaying of the score still uses the integer using songScoreInt.
Written with the assumption that https://github.com/FunkinCrew/Funkin/pull/3708 will be pulled. If it isn't, removals of offset corrections will be required.

xenkap committed:
Consistent sustains + songScore to Float

xenkap commented:
> An alternative approach would be to add a variable in SustainTrail.hx that tracks PlayState's last read on it, and then performing logic where it used to be done anyway. I'll might make a separate branch for that later-- if I don't, well.
On second thought, this seems like the simpler approach. Working on it now

Hundrec commented:
If done right, this could finally standardize scores across all attempts!
Please polish this up the best you can :)

xenkap committed:
Separate sustain rewards from isPlayer variable

Modders may want to modify sustain/input behavior, like in custom input engines or opponent mode scripts. Or other things.

xenkap commented:
> If done right, this could finally standardize scores across all attempts! Please polish this up the best you can :)

Thank you!! I'm not sure what much else I can do on the topic of sustains, since it seems to be 100% accurate now.
But talking about standardized scores, I'm not very sure about the 5ms input window for a perfect score, haha
Although that might be reasonable in other rhythm games. Haven't ever seen one with that strict of a window, though

xenkap commented:
I just realized the after video shows me taking a score of 88.
I've already fixed this, but since I recorded it before that, it doesn't appear so. Bwekh

VioletSnowLeopard committed:
Update CHANGELOG.md

VioletSnowLeopard committed:
put fixed back

xenkap committed:
MS based release !!!!

honk honk. i have come to deliver more JANK

xenkap commented:
Rebased, fixed merge conflicts, squashed it a bit to make it cleaner :fire:
Hoping for a review once more! I think the PR has what it needs, but if there's improvement or changes to be made, then :eyes:

Hundrec commented:
Thanks! Would still love to see this become a consistent standard for scoring

xenkap commented:
Nevermind, looks like it should still work fine
I'm not the biggest fan of having scoring be handled by Strumline still, ~but I can't see much of a way around it :sob:~
Scratch that, I do have an idea -- while it won't really fix it, changing out the PlayState.instance.sustainHit calls with a callback instead would allow it to be unspecified or not be strictly tied to PlayState. I think that'd be better !!

on it :boom:

xenkap committed:
onSustainHit callback

h

Hundrec commented:
Aw darn, more conflicts

FuroYT commented:
We need you for SDL3 vro
<img width="434" height="670" alt="sentpajim" src="https://github.com/user-attachments/assets/60c562ba-2c64-4b9d-a316-177af933db87" />
